src/web_analysis/test_robots.py

```diff
@@ -0,0 +1,1087 @@
+        )
+
+    def test_empty_robots_txt(self):
+        """Tests behavior when the robots.txt is empty, implying all access should be allowed."""
+        agent_rules = {}
+        all_agents = ["GoogleBot"]
+        expected = {"*": "none", "GoogleBot": "none"}
+        self.assertEqual(
+            interpret_robots(agent_rules, all_agents),
+            expected,
+            "An empty robots.txt should allow all access, but it does not.",
+        )
+
+    def test_precedence_of_directives(self):
+        """Tests precedence of Allow over Disallow directives within the same robots.txt."""
+        agent_rules = {"GoogleBot": {"Disallow": ["/"], "Allow": ["/public"]}}
+        all_agents = ["GoogleBot"]
+        expected = {"*": "none", "GoogleBot": "some"}
+        self.assertEqual(
+            interpret_robots(agent_rules, all_agents),
+            expected,
+            "Allow directives should override conflicting Disallow directives, but they do not.",
+        )
+
+    def test_disallow_all_but_one(self):
+        """Tests disallowing all paths except one specific allowed path."""
+        agent_rules = {"*": {"Disallow": ["/"], "Allow": ["/index.html"]}}
+        all_agents = ["GoogleBot", "BingBot"]
+        expected = {"*": "all", "GoogleBot": "some", "BingBot": "some"}
+        self.assertEqual(
+            interpret_robots(agent_rules, all_agents),
+            expected,
+            "Disallowing all except one file should block all but that file. Check for incorrect rule application.",
+        )
+
+    def test_case_insensitive_paths(self):
+        """Tests that the path directives are case insensitive."""
+        agent_rules = {"*": {"Disallow": ["/Folder"]}}
+        all_agents = ["GoogleBot"]
+        expected = {"*": "some", "GoogleBot": "some"}
+        self.assertEqual(
+            interpret_robots(agent_rules, all_agents),
+            expected,
+            "Path directives should be case insensitive, ensuring '/Folder' matches '/folder'.",
+        )
+
+    def test_query_parameters_in_paths(self):
+        """Tests that paths including query parameters are handled correctly."""
+        agent_rules = {"*": {"Disallow": ["/search?query=public"]}}
+        all_agents = ["GoogleBot"]
+        expected = {"*": "some", "GoogleBot": "some"}
+        self.assertEqual(
+            interpret_robots(agent_rules, all_agents),
+            expected,
+            "Paths with query parameters should be respected in disallow rules.",
+        )
+
+    def test_utf8_path_encodings(self):
+        """Tests the handling of UTF-8 encoded paths."""
+        agent_rules = {"*": {"Disallow": ["/résumé"]}}
+        all_agents = ["GoogleBot"]
+        expected = {"*": "some", "GoogleBot": "some"}
+        self.assertEqual(
+            interpret_robots(agent_rules, all_agents),
+            expected,
+            "UTF-8 paths should be correctly interpreted and matched.",
+        )
+
+    def test_empty_user_agent_disallow(self):
```

Comment:
I will add a check to make sure if the agent is empty it won't be processed. So the outcome of this test should never happen